### ISSUE TITLE
Allow accessing enum members by enum name

### ISSUE THREAD

**Issue by Kixiron** [vmware/differential-datalog#810]
Large programs can often involve very large and very numerous enums and with the current design all enum variants are simultaneously in scope, which leads to name conflicts, having to rename variants, etc.

A system to allow referring to variants by their enum's name would alleviate this, e.g.

```prolog
typedef Enum = Variant1 | Variant2

Enum.Variant1
// or
Enum::Variant2
```

Backwards compatibility is a concern, this could be solved by making a separate enum declaration, making named access a compiler flag or something else entirely

**Comment by ryzhyk** [comment]
The `::` syntax is probably cleaner, as it is generally used as a static namespace separator in DDlog, whereas `.` accesses a field or method of an object.

In DDlog there is no difference between structs and enums.  The former is just an enum with a unique constructor.  However, it would not be nice to force users to always type both struct name and constructor name, which are often the same. Even for proper enums, typing fully qualified names can be an overkill in some cases, especially for ubiquitous things like `Option::Some`.  So perhaps the `::` syntax can be optional and only required when necessary to avoid ambiguity.

Rust allows explicit control over whether enumerator names are visible within their own namespace or directly in the module namespace. This seems like a good design (like so many other things in Rust) and ultimately the right thing to do, but it would make the DDlog module system significantly more complex.

> Backwards compatibility is a concern, this could be solved by making a separate enum declaration,

This sounds hacky to me.

> making named access a compiler flag 

That's doable, but we can support both modes at the same time as described above.

**Comment by Kixiron** [comment]
Let me elaborate on my thoughts a wee bit:

For the enum thing, this'd be an entirely new declaration type, so it's distinct (visually) and doesn't break or even touch backwards compatibility

```prolog
enum Foo = Bar | Baz
```

Alternatively, an attribute could do more or less the same thing

```prolog
#[enum]
typedef Foo = Bar | Baz
```

As for fixing having to constantly type `Option::Some`, I think slightly increasing the complexity of import statements could help with that and some other quality of life stuff

```prolog
import ddlog_std exposing { Option, Option::Some, Option::None, Either }
```

**Comment by ryzhyk** [comment]
> For the enum thing, this'd be an entirely new declaration type, so it's distinct (visually) and doesn't break or even touch backwards compatibility

Right, this is how I understood your original suggestion, but I don't think it's a good idea to add a new construct to the language (and thus make the language and the compiler more complex) just to preserve backwards compatibility. I'd rather break things but keep the language clean.

> Alternatively, an attribute could do more or less the same thing

Attributes are not supposed to change the semantics or syntax of the language.  I think this would not be a valid use of an attribute.

> As for fixing having to constantly type Option::Some, I think slightly increasing the complexity of import statements could help with that and some other quality of life stuff

True, but to make this really useful, we'd also have to add features to re-export declarations, control external visibility of declarations, etc.  This is probably the way to go, and maybe we should shelve this until we have the cycles for a major overhaul of DDlog's fairly rudimentary module system.
